File: java/3479.0_Fruits_Into_Baskets_III.java
class SegmentTree {
    private final int[] max;

    public SegmentTree(int[] a) {
        int n = a.length;
        max = new int[2 << (32 - Integer.numberOfLeadingZeros(n - 1))];
        build(a, 1, 0, n - 1);
    }

    public int findFirstAndUpdate(int o, int l, int r, int x) {
        if (max[o] < x) {
            return -1;
        }
        if (l == r) {
            max[o] = -1;
            return l;
        }
        int m = (l + r) / 2;
        int i = findFirstAndUpdate(o * 2, l, m, x);
        if (i < 0) {
            i = findFirstAndUpdate(o * 2 + 1, m + 1, r, x);
        }
        maintain(o);
        return i;
    }

    private void maintain(int o) {
        max[o] = Math.max(max[o * 2], max[o * 2 + 1]);
    }

    private void build(int[] a, int o, int l, int r) {
        if (l == r) {
            max[o] = a[l];
            return;
        }
        int m = (l + r) / 2;
        build(a, o * 2, l, m);
        build(a, o * 2 + 1, m + 1, r);
        maintain(o);
    }
}

class Solution {
    public int numOfUnplacedFruits(int[] fruits, int[] baskets) {
        SegmentTree t = new SegmentTree(baskets);
        int n = baskets.length;
        int ans = 0;
        for (int x: fruits) {
            if (t.findFirstAndUpdate(1, 0, n - 1, x) < 0) {
                ans++;
            }
        }
        return ans;
    }
}

File: go/3479.0_Fruits_Into_Baskets_III.go
type seg []int

func (t seg) maintain(o int) {
    t[o] = max(t[o<<1], t[o<<1|1])
}

func (t seg) build(a []int, o, l, r int) {
    if l == r {
        t[o] = a[l];
        return
    }
    m := (l + r) >> 1
    t.build(a, o<<1, l, m)
    t.build(a, o <<1|1, m + 1, r)
    t.maintain(o)
}

func (t seg) findFirstAndUpdate(o, l, r, x int) int {
    if t[o] < x {
        return -1
    }
    if l == r {
        t[o] = -1
        return l
    }
    m := (l + r) >> 1
    i := t.findFirstAndUpdate(o<<1, l, m, x)
    if i < 0 {
        i = t.findFirstAndUpdate(o<<1|1, m+1, r, x)
    }
    t.maintain(o)
    return i
}

func newSegmentTree(a []int) seg {
    n := len(a)
    t := make(seg, 2<<bits.Len(uint(n-1)))
    t.build(a, 1, 0, n - 1)
    return t
}

func numOfUnplacedFruits(fruits []int, baskets []int) (ans int) {
    t := newSegmentTree(baskets)
    for _, x := range fruits {
        if t.findFirstAndUpdate(1, 0, len(baskets)-1, x) < 0 {
            ans++
        }
    }
    return 
}